zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -314,7 +320,44 @@ where
 
             let miner_fee = miner_fee(&mempool_txs);
 
+            // Calling state with `ChainInfo` request for relevant chain data
+            let request = ReadRequest::ChainInfo;
+            let response = state
+                .ready()
+                .and_then(|service| service.call(request))
+                .await
+                .map_err(|error| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                                data: None,
+                            })?;
+
+            let chain_info = match response {
+                ReadResponse::ChainInfo(Some(chain_info)) => chain_info,
+                _ => unreachable!("lets hope for always some until later"),
+            };
+
+            // Get the tip data from the state call
+            let tip_height = chain_info.tip.0;
+            let tip_hash = chain_info.tip.1;
+
             let block_height = (tip_height + 1).expect("tip is far below Height::MAX");
+            let min_time = chain_info.median_time_past.checked_add_signed(Duration::seconds(1))
+                .expect("median time plus a small constant is far below i64::MAX");
+
+            // > For each block at block height 2 or greater on Mainnet, or block height 653606 or greater on Testnet, nTime
+            // > MUST be less than or equal to the median-time-past of that block plus 90 * 60 seconds.
+            //
+            // We ignore the height as we are checkpointing on nu5 in Mainnet and Testnet.
+            let cur_time = chain_info.current_system_time;
+            if cur_time > min_time.checked_add_signed(Duration::seconds(5400)).expect("median time plus a small constant is far below i64::MAX") {
+                return Err(Error {
+                    code: ErrorCode::ServerError(0),
+                    message: format!("Current time {cur_time} is bigger than {min_time} + 5400 seconds."),
+                    data: None,
+                });
+            }
```

Comment:
These checks are tricky to get right, what do you think of these changes?

In the suggestion below:
1. The `max_time` calculation seems like it is off by one - the rule is:
> less than or equal to the median-time-past of that block plus 90 * 60 seconds
But this code calculates `median_time_past + 1 + 90*60 seconds`, so any user generating blocks with the largest possible time will create invalid blocks.

2. The `cur_time` must be greater than or equal to `min_time`, but this code does not check that.

3. If we return an error when `cur_time` is greater than `max_time`, users with bad clock skew won't be able to mine blocks. And if mainnet or testnet go down for 90 minutes, all Zebra instances won't be able to mine blocks. Instead, we can clamp the time so it is in the correct range.

Other changes:

After this, the time is correct for mainnet, but testnet has an extra consensus rule that can change the difficulty based on the time:
> if the block time of a block at height height ≥ 299188 is greater than 6 · PoWTargetSpacing(height) seconds after that of the preceding block, then the block is a minimum-difficulty block. In that case its nBits field MUST be set to ToCompact(PoWLimit)

https://zips.z.cash/zip-0208#minimum-difficulty-blocks-on-testnet

4. This check can change the `cur_time`, so we need to move this whole check into the state request, before we call `AdjustedDifficulty::new_from_header_time()`. Otherwise, we could be using a one time to calculate the difficulty, and a different time in the block header.

5. The miner can also change the time to anything between `min_time` and `max_time`. But on testnet, changing the time can change the difficulty. So if `max_time` is a minimum difficulty block, we can either:
- remove "time" from `getblocktemplate.mutable`
- change `min_time` and `cur_time` to the first time that allows a minimum difficulty block, and call `AdjustedDifficulty::new_from_header_time()` with the new `cur_time`, so all times the miner can use are minimum difficulty blocks

What do you think we should do here?

I don't really mind either way, but changing `min_time` and `cur_time` might make testing faster, because some blocks will be easier. 

```suggestion
            let max_time = chain_info.median_time_past.checked_add_signed(Duration::seconds(5400)).expect("median time plus a small constant is far below i64::MAX");
            let cur_time = chain_info.current_system_time.timestamp().clamp(min_time.timestamp(), max_time.timestamp());
       }
```

Reply:
Lets follow up in new comments.

Reply:
> The miner can also change the time to anything between min_time and max_time. But on testnet, changing the time can change the difficulty. So if max_time is a minimum difficulty block, we can either:
>
> - remove "time" from getblocktemplate.mutable
> - change min_time and cur_time to the first time that allows a minimum difficulty block, and call AdjustedDifficulty::new_from_header_time() with the new cur_time, so all times the miner can use are minimum difficulty blocks
>
> What do you think we should do here?
>
> I don't really mind either way, but changing min_time and cur_time might make testing faster, because some blocks will be easier.

This is still missing.

Reply:
> This is still missing.

I added some code for the remove time option in https://github.com/ZcashFoundation/zebra/pull/5659/commits/3a862aa942e24f0f28024f307358c7f32f6b128d